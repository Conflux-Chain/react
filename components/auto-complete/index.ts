import AutoComplete, { useAutoCompleteHandle } from './auto-complete'
export { useAutoCompleteHandle }
export default AutoComplete
