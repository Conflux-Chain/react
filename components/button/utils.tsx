import React, { ReactNode } from 'react'
import { NormalSizes } from '../utils/prop-types'
import ButtonIcon from './button-icon'
import { ButtonProps } from './button'
import { ButtonGroupConfig } from '../button-group/button-group-context'

export const getButtonChildrenWithIcon = (
  loading: boolean,
  auto: boolean,
  size: NormalSizes,
  children: ReactNode,
  icons: {
    icon?: React.ReactNode
    iconRight?: React.ReactNode
  },
  isTag: boolean,
) => {
  const { icon, iconRight } = icons
  const hasIcon = icon || iconRight
  const isRight = Boolean(iconRight)
  const paddingForAutoMode =
    auto || size === 'mini' || isTag
      ? `calc(var(--zeit-ui-button-height) / 2 + var(--zeit-ui-button-padding) * .5)`
      : 0
  if (!hasIcon) return <div className={`text ${loading ? 'hidden' : ''}`}>{children}</div>
  return (
    <>
      <ButtonIcon isRight={isRight}>{hasIcon}</ButtonIcon>
      <div className={`text ${isRight ? 'right' : 'left'} ${loading ? 'hidden' : ''}`}>
        {children}
        <style jsx>{`
          .left {
            padding-left: ${paddingForAutoMode};
          }
          .right {
            padding-right: ${paddingForAutoMode};
          }
        `}</style>
      </div>
    </>
  )
}

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig,
): ButtonProps => {
  if (!config.isButtonGroup) return props
  return {
    ...props,
    auto: true,
    shadow: false,
    ghost: config.ghost || props.ghost,
    size: config.size || props.size,
    variant: config.variant || props.variant,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  }
}
