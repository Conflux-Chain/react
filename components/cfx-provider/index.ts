import CfxProvider from './cfx-provider'

export default CfxProvider
