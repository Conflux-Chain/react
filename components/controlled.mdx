If you think the data flow in react carefully.

uncontrolled component:
caller do not know the internal state of the component

controlled component:
the state of called component is fully controlled by `props` passed in 

The above two are identified by react offical team. However, there is also something 
in between, which can not fit in the patterns mentioned above

third patten: the called need to know the callee state sometims but not always
think form as an example, the caller only need to know the state of callee when 'submit'
happen rather than keep track of the component all the time

forth pattern: the caller need to set the callee state forcelly on a one time, not inter
