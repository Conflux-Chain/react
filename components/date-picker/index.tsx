import { Moment } from 'moment'
import momentGenerateConfig from '@jnoodle/rc-picker/lib/generate/moment'
import generatePicker, {
  PickerProps,
  PickerDateProps,
  RangePickerProps as BaseRangePickerProps,
} from './generate-picker'

export type DatePickerProps = PickerProps<Moment>
export type MonthPickerProps = Omit<PickerDateProps<Moment>, 'picker'>
export type WeekPickerProps = Omit<PickerDateProps<Moment>, 'picker'>
export type RangePickerProps = BaseRangePickerProps<Moment>

export const DatePicker = generatePicker<Moment>(momentGenerateConfig)

export default DatePicker
