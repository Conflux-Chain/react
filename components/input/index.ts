import Input, { useInputHandle } from './input'

export { useInputHandle }
export default Input
