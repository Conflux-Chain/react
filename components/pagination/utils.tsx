export const getPageCount = (total: number, pageSize: number) => {
  return Math.floor((total - 1) / pageSize) + 1
}
