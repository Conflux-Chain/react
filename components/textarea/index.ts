import Textarea, { useTextareaHandle } from './textarea'

Textarea.useTextareaHandle = useTextareaHandle

export default Textarea
export { useTextareaHandle }
