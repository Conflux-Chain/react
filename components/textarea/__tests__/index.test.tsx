import { Textarea } from 'components'
import { mount } from 'enzyme'
import React from 'react'
import { nativeEvent } from 'tests/utils'

describe('Textarea', () => {
  it('should render correctly', () => {
    const wrapper = mount(<Textarea placeholder="placeholder" />)
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should work with different styles', () => {
    const wrapper = mount(
      <div>
        <Textarea status="primary" />
        <Textarea width="20%" />
        <Textarea minHeight="100px" />
        <Textarea solid />
      </div>,
    )
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => wrapper.unmount()).not.toThrow()
  })

  it('should set textarea from value', () => {
    const wrapper = mount(<Textarea initialValue="test-value" />)
    let el = wrapper.find('textarea').getDOMNode() as HTMLTextAreaElement
    expect(el.value).toEqual('test-value')

    wrapper.setProps({ value: 'test-value2' })
    el = wrapper.find('textarea').getDOMNode() as HTMLTextAreaElement
    expect(el.value).toEqual('test-value2')
  })

  it('should trigger events when textarea changed', () => {
    let value = ''
    const handler = jest.fn().mockImplementation(e => (value = e.target.value))
    const wrapper = mount(<Textarea onChange={handler} />)
    wrapper.find('textarea').simulate('change', { target: { value: 'test-value' } })
    expect(handler).toHaveBeenCalled()
    expect(value).toEqual('test-value')
    handler.mockRestore()
  })

  it('should ignore events when disabled or readonly', () => {
    const handler = jest.fn()
    const wrapper = mount(<Textarea onChange={handler} disabled />)
    wrapper.find('textarea').simulate('change', { target: { value: 'test-value' } })
    expect(handler).not.toHaveBeenCalled()

    wrapper.setProps({ disabled: false, readOnly: true })
    wrapper.find('textarea').simulate('change', { target: { value: 'test-value2' } })
    expect(handler).not.toHaveBeenCalled()
    handler.mockRestore()
  })

  it('should pass through blur event', () => {
    const blurHandler = jest.fn()
    const focusHandler = jest.fn()
    const wrapper = mount(<Textarea onBlur={blurHandler} onFocus={focusHandler} />)

    wrapper.find('textarea').simulate('focus', nativeEvent)
    expect(focusHandler).toHaveBeenCalled()

    wrapper.find('textarea').simulate('blur', nativeEvent)
    expect(blurHandler).toHaveBeenCalled()
  })

  it('should pass through onMouseOut event', () => {
    const onMouseOver = jest.fn()
    const onMouseOut = jest.fn()
    const wrapper = mount(<Textarea onMouseOver={onMouseOver} onMouseOut={onMouseOut} />)

    wrapper.find('textarea').simulate('mouseOver', nativeEvent)
    expect(onMouseOver).toHaveBeenCalled()

    wrapper.find('textarea').simulate('mouseOut', nativeEvent)
    expect(onMouseOut).toHaveBeenCalled()
  })

  it('should show the right character count', () => {
    const wrapper = mount(<Textarea counter initialValue="foo" />)
    expect(wrapper.exists('.counter .separator')).toBeFalsy
    expect(wrapper.exists('.counter .limit')).toBeFalsy
    expect(wrapper.find('.counter .count').text()).toBe('3')
  })

  it("should limit and slice user's input if reaches the maxLength limit", () => {
    const onChange = jest.fn()
    const wrapper = mount(<Textarea counter maxLength={3} initialValue="f" onChange={onChange} />)
    let changeEvent = { target: { value: 'fo' } }
    wrapper.find('textarea').simulate('change', changeEvent)
    expect(onChange).toHaveBeenCalledWith(expect.objectContaining(changeEvent))
    changeEvent = { target: { value: 'foo' } }
    wrapper.find('textarea').simulate('change', changeEvent)
    expect(onChange).toHaveBeenCalledWith(expect.objectContaining(changeEvent))
    changeEvent = { target: { value: 'fooo' } }
    wrapper.find('textarea').simulate('change', changeEvent)
    expect(onChange).toHaveBeenCalledWith(expect.objectContaining({ target: { value: 'foo' } }))
  })

  it('should forward ref by default', () => {
    const ref = React.createRef<HTMLTextAreaElement>()
    const wrapper = mount(<Textarea ref={ref} />)
    expect(ref.current).not.toBeNull()
    expect(() => wrapper.unmount()).not.toThrow()
  })
})
