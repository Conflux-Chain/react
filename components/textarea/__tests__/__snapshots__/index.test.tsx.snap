// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea should render correctly 1`] = `
"<div class=\\"wrapper   lined  \\"><textarea placeholder=\\"placeholder\\"></textarea><style>
          .wrapper {
            position: inherit;
            display: inline-flex;
            box-sizing: border-box;
            user-select: none;
            width: initial;
            min-width: 12.5rem;
            max-width: 95vw;
            height: auto;
            border-radius: 4px;
            color: #282d30;
            transition: border 0.2s ease 0s, color 0.2s ease 0s;
            background-color: #fdfdfe;
            border: 1px solid #ebeced;
          }

          .wrapper.hover:not(.disabled),
          .wrapper.focus:not(.disabled) {
            border-color: #0054fe;
            background-color: #fdfdfe;
          }

          .wrapper.disabled {
            color: #ebeced;
            border-color: #ebeced;
            cursor: not-allowed;
            opacity: 0.5;
          }

          textarea {
            background-color: transparent;
            box-shadow: none;
            display: block;
            font-family: \\"Circular Std Book\\", \\"PingFang SC\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", sans-serif;
            font-size: 0.875rem;
            width: 100%;
            height: 100%;
            min-height: 6.25rem;
            resize: none;
            border: none;
            outline: none;
            padding: calc(16px * 0.875) calc(16px * 0.75);
          }

          textarea:disabled {
            cursor: not-allowed;
          }

          ::placeholder,
          ::-moz-placeholder,
          :-ms-input-placeholder,
          ::-webkit-textarea-placeholder {
            color: #9b9eac;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active,
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-background-clip: text;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active {
            -webkit-box-shadow: 0 0 0 30px #fff inset !important;
          }
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-box-shadow: 0 0 0 30px #fdfdfe inset !important;
          }
        </style></div>"
`;

exports[`Textarea should work with different styles 1`] = `
"<div><div class=\\"wrapper   lined  \\"><textarea></textarea><style>
          .wrapper {
            position: inherit;
            display: inline-flex;
            box-sizing: border-box;
            user-select: none;
            width: initial;
            min-width: 12.5rem;
            max-width: 95vw;
            height: auto;
            border-radius: 4px;
            color: #282d30;
            transition: border 0.2s ease 0s, color 0.2s ease 0s;
            background-color: #fdfdfe;
            border: 1px solid #c0d8fc;
          }

          .wrapper.hover:not(.disabled),
          .wrapper.focus:not(.disabled) {
            border-color: #0054fe;
            background-color: #fdfdfe;
          }

          .wrapper.disabled {
            color: #ebeced;
            border-color: #ebeced;
            cursor: not-allowed;
            opacity: 0.5;
          }

          textarea {
            background-color: transparent;
            box-shadow: none;
            display: block;
            font-family: \\"Circular Std Book\\", \\"PingFang SC\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", sans-serif;
            font-size: 0.875rem;
            width: 100%;
            height: 100%;
            min-height: 6.25rem;
            resize: none;
            border: none;
            outline: none;
            padding: calc(16px * 0.875) calc(16px * 0.75);
          }

          textarea:disabled {
            cursor: not-allowed;
          }

          ::placeholder,
          ::-moz-placeholder,
          :-ms-input-placeholder,
          ::-webkit-textarea-placeholder {
            color: #9b9eac;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active,
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-background-clip: text;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active {
            -webkit-box-shadow: 0 0 0 30px #fff inset !important;
          }
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-box-shadow: 0 0 0 30px #fdfdfe inset !important;
          }
        </style></div><div class=\\"wrapper   lined  \\"><textarea></textarea><style>
          .wrapper {
            position: inherit;
            display: inline-flex;
            box-sizing: border-box;
            user-select: none;
            width: 20%;
            min-width: 12.5rem;
            max-width: 95vw;
            height: auto;
            border-radius: 4px;
            color: #282d30;
            transition: border 0.2s ease 0s, color 0.2s ease 0s;
            background-color: #fdfdfe;
            border: 1px solid #ebeced;
          }

          .wrapper.hover:not(.disabled),
          .wrapper.focus:not(.disabled) {
            border-color: #0054fe;
            background-color: #fdfdfe;
          }

          .wrapper.disabled {
            color: #ebeced;
            border-color: #ebeced;
            cursor: not-allowed;
            opacity: 0.5;
          }

          textarea {
            background-color: transparent;
            box-shadow: none;
            display: block;
            font-family: \\"Circular Std Book\\", \\"PingFang SC\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", sans-serif;
            font-size: 0.875rem;
            width: 100%;
            height: 100%;
            min-height: 6.25rem;
            resize: none;
            border: none;
            outline: none;
            padding: calc(16px * 0.875) calc(16px * 0.75);
          }

          textarea:disabled {
            cursor: not-allowed;
          }

          ::placeholder,
          ::-moz-placeholder,
          :-ms-input-placeholder,
          ::-webkit-textarea-placeholder {
            color: #9b9eac;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active,
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-background-clip: text;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active {
            -webkit-box-shadow: 0 0 0 30px #fff inset !important;
          }
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-box-shadow: 0 0 0 30px #fdfdfe inset !important;
          }
        </style></div><div class=\\"wrapper   lined  \\"><textarea></textarea><style>
          .wrapper {
            position: inherit;
            display: inline-flex;
            box-sizing: border-box;
            user-select: none;
            width: initial;
            min-width: 12.5rem;
            max-width: 95vw;
            height: auto;
            border-radius: 4px;
            color: #282d30;
            transition: border 0.2s ease 0s, color 0.2s ease 0s;
            background-color: #fdfdfe;
            border: 1px solid #ebeced;
          }

          .wrapper.hover:not(.disabled),
          .wrapper.focus:not(.disabled) {
            border-color: #0054fe;
            background-color: #fdfdfe;
          }

          .wrapper.disabled {
            color: #ebeced;
            border-color: #ebeced;
            cursor: not-allowed;
            opacity: 0.5;
          }

          textarea {
            background-color: transparent;
            box-shadow: none;
            display: block;
            font-family: \\"Circular Std Book\\", \\"PingFang SC\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", sans-serif;
            font-size: 0.875rem;
            width: 100%;
            height: 100%;
            min-height: 100px;
            resize: none;
            border: none;
            outline: none;
            padding: calc(16px * 0.875) calc(16px * 0.75);
          }

          textarea:disabled {
            cursor: not-allowed;
          }

          ::placeholder,
          ::-moz-placeholder,
          :-ms-input-placeholder,
          ::-webkit-textarea-placeholder {
            color: #9b9eac;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active,
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-background-clip: text;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active {
            -webkit-box-shadow: 0 0 0 30px #fff inset !important;
          }
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-box-shadow: 0 0 0 30px #fdfdfe inset !important;
          }
        </style></div><div class=\\"wrapper   solid  \\"><textarea></textarea><style>
          .wrapper {
            position: inherit;
            display: inline-flex;
            box-sizing: border-box;
            user-select: none;
            width: initial;
            min-width: 12.5rem;
            max-width: 95vw;
            height: auto;
            border-radius: 4px;
            color: #282d30;
            transition: border 0.2s ease 0s, color 0.2s ease 0s;
            background-color: #f9fafb;
            border: transparent;
          }

          .wrapper.hover:not(.disabled),
          .wrapper.focus:not(.disabled) {
            border-color: transparent;
            background-color: #e8f4ff;
          }

          .wrapper.disabled {
            color: #ebeced;
            border-color: #ebeced;
            cursor: not-allowed;
            opacity: 0.5;
          }

          textarea {
            background-color: transparent;
            box-shadow: none;
            display: block;
            font-family: \\"Circular Std Book\\", \\"PingFang SC\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", \\"Roboto\\", \\"Oxygen\\", \\"Ubuntu\\", \\"Cantarell\\", \\"Fira Sans\\", \\"Droid Sans\\", \\"Helvetica Neue\\", sans-serif;
            font-size: 0.875rem;
            width: 100%;
            height: 100%;
            min-height: 6.25rem;
            resize: none;
            border: none;
            outline: none;
            padding: calc(16px * 0.875) calc(16px * 0.75);
          }

          textarea:disabled {
            cursor: not-allowed;
          }

          ::placeholder,
          ::-moz-placeholder,
          :-ms-input-placeholder,
          ::-webkit-textarea-placeholder {
            color: #9b9eac;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active,
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-background-clip: text;
          }

          textarea:-webkit-autofill,
          textarea:-webkit-autofill:active {
            -webkit-box-shadow: 0 0 0 30px #fff inset !important;
          }
          textarea:-webkit-autofill:hover,
          textarea:-webkit-autofill:focus {
            -webkit-box-shadow: 0 0 0 30px #e8f4ff inset !important;
          }
        </style></div></div>"
`;
